src/repr/src/adt/timestamp.rs

```diff
@@ -50,6 +54,94 @@ const NANOSECONDS_PER_HOUR: i64 = NANOSECONDS_PER_MINUTE * MINUTES_PER_HOUR;
 const NANOSECONDS_PER_MINUTE: i64 = NANOSECONDS_PER_SECOND * SECONDS_PER_MINUTE;
 const NANOSECONDS_PER_SECOND: i64 = 10i64.pow(9);
 
+pub const MAX_PRECISION: u8 = 6;
+
+/// The `max_precision` of a [`ScalarType::Timestamp`] or
+/// [`ScalarType::TimestampTz`].
+///
+/// This newtype wrapper ensures that the length is within the valid range.
+///
+/// [`ScalarType::Timestamp`]: crate::ScalarType::Timestamp
+/// [`ScalarType::TimestampTz`]: crate::ScalarType::TimestampTz
+#[derive(
+    Arbitrary,
+    Debug,
+    Clone,
+    Copy,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    Serialize,
+    Deserialize,
+    MzReflect,
+)]
+pub struct TimestampPrecision(pub(crate) u8);
+
+impl TimestampPrecision {
+    /// Consumes the newtype wrapper, returning the inner `u8`.
+    pub fn into_u8(self) -> u8 {
+        self.0
+    }
+}
+
+impl TryFrom<i64> for TimestampPrecision {
+    type Error = InvalidTimestampPrecisionError;
+
+    fn try_from(max_precision: i64) -> Result<Self, Self::Error> {
+        match u8::try_from(max_precision) {
+            Ok(max_precision) if max_precision <= MAX_PRECISION => {
+                Ok(TimestampPrecision(max_precision))
+            }
+            _ => Err(InvalidTimestampPrecisionError),
+        }
+    }
+}
+
+impl RustType<ProtoTimestampPrecision> for TimestampPrecision {
+    fn into_proto(&self) -> ProtoTimestampPrecision {
+        ProtoTimestampPrecision {
+            value: self.0.into_proto(),
+        }
+    }
+
+    fn from_proto(proto: ProtoTimestampPrecision) -> Result<Self, TryFromProtoError> {
+        Ok(TimestampPrecision(proto.value.into_rust()?))
+    }
+}
+
+impl RustType<ProtoOptionalTimestampPrecision> for Option<TimestampPrecision> {
+    fn into_proto(&self) -> ProtoOptionalTimestampPrecision {
+        ProtoOptionalTimestampPrecision {
+            value: self.into_proto(),
+        }
+    }
+
+    fn from_proto(precision: ProtoOptionalTimestampPrecision) -> Result<Self, TryFromProtoError> {
+        precision.value.into_rust()
+    }
+}
+
+/// The error returned when constructing a [`VarCharMaxLength`] from an invalid
+/// value.
+///
+/// [`VarCharMaxLength`]: crate::adt::varchar::VarCharMaxLength
+#[derive(Debug, Clone)]
+pub struct InvalidTimestampPrecisionError;
+
+impl fmt::Display for InvalidTimestampPrecisionError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "precision for type timestamp or timestamptz must be between 1 and {}",
```

Comment:
There's no check currently for the minimum precision to be 1. As it's implemented, a precision of 0 will be equivalent to `seconds` precision (i.e. no fractional bits). Does it make sense to keep 0 as the minimum? or should it be changed to 1 (will add a check then)?

Reply:
Oh sorry, forgot to fix the error message. Yes, to match PostgreSQL we should support a precision of 0. (Unlike numeric types, where zero precision is useless, a zero-precision timestamp is still useful, as it just means you don’t want any fractional seconds.)

Reply:
Agree, makes sense.